hera_pspec/pspecbeam.py

```diff
@@ -0,0 +1,291 @@
+
+        little_h : boolean, optional
+                Whether to have cosmological length units be h^-1 Mpc or Mpc
+                Default: h^-1 Mpc
+
+        num_steps : int, optional
+                Number of steps to use when interpolating primary beams for
+                numerical integral
+                Default: 10000
+
+        Returns
+        -------
+        scalar: float
+                [\int dnu (\Omega_PP / \Omega_P^2) ( B_PP / B_P^2 ) / (X^2 Y)]^-1
+                in h^-3 Mpc^3 or Mpc^3.
+        """
+
+        integration_freqs = np.linspace(lower_freq,upper_freq,num_steps)
+        integration_freqs_MHz = integration_freqs / 10**6 # The interpolations are generally more stable in MHz
+        redshifts = conversion.f2z(integration_freqs).flatten()
+        X2Y = np.array(map(conversion.X2Y,redshifts))
+
+        # Use linear interpolation to interpolate the frequency-dependent quantities
+        # derived from the beam model to the same frequency grid as the power spectrum
+        # estimation
+        beam_model_freqs_MHz = self.beam_freqs/(1.*10**6)
+        dOpp_over_Op2_fit = interp1d(beam_model_freqs_MHz, self.power_beam_sq_int(stokes)/self.power_beam_int(stokes)**2)
+        dOpp_over_Op2 = dOpp_over_Op2_fit(integration_freqs_MHz)
+
+
+        # Get B_pp = \int dnu taper^2 and Bp = \int dnu
+        if taper == 'none':
+            dBpp_over_BpSq = np.ones_like(integration_freqs)
+        else:
+            dBpp_over_BpSq = aipy.dsp.gen_window(num_freqs,taper)**2
+        dBpp_over_BpSq /= (integration_freqs[-1] - integration_freqs[0])**2
+
+        d_inv_scalar = dBpp_over_BpSq * dOpp_over_Op2 /  X2Y
+
+        scalar = 1 / integrate.trapz(d_inv_scalar, integration_freqs)
+        if little_h == True:
+            scalar *= conversion.h**3
+
+        return scalar
+
+
+
+class PSpecBeamUV(PSpecBeam):
+
+    def __init__(self, beam_fname):
+        """
+        Object to store the primary beam for a pspec observation.
+        This is subclassed from PSpecBeam to take in a pyuvdata
+        UVBeam object.
+
+        Parameters
+        ----------
+        beam_fname: str
+                Path to a pyuvdata UVBeam file
+        """
+        self.primary_beam = pyuvdata.UVBeam()
+        self.primary_beam.read_beamfits(beam_fname)
+        self.beam_freqs = self.primary_beam.freq_array[0]
+
+    def power_beam_int(self, stokes='I'):
+        """
+        Computes the integral of the beam over solid angle to give
+        a beam area (in str) as a function of frequency. Uses function
+        in pyuvdata.
+
+        Parameters
+        ----------
+        stokes: str, optional
+                Which Stokes parameter's beam to compute the scalar for.
+                'I', 'Q', 'U', 'V', although currently only 'I' is implemented
+                Default: 'I'
+        Returns
+        -------
+        primary beam area: float, array-like
+        """
+        return self.primary_beam.get_beam_area(stokes)
```

Comment:
For now, can we protect this call in a `try`/`except` structure? This will help with backwards compatibility etc. while we're waiting for the changes to pyuvdata to be made (and for users to update their pyuvdata installations).

Reply:
Done!